### PR TITLE
Implementation of coscheduling based on CRD [Part 1]

### PR DESCRIPTION
**Pull request by cwdsuzhou**
Part 1: we implement the controller and crd
.
/cc @Huang-Wei

### CODE REVIEW

**Comment by denkensk** [comment]
We need to separate this PR to two @cwdsuzhou @Huang-Wei 
1. CRD Define + Controller
2. Use the PodGroup in the current coscheduling implemention (instead of deleteing the current coscheduling directly)

File: pkg/coscheduling/apis/podgroup/v1alpha1/doc.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/*
```
Comment (denkensk) [inline pkg/coscheduling/apis/podgroup/v1alpha1/doc.go:1]: Move the apis from  pkg/coscheduling/apis to pkg/apis/

Comment (cwdsuzhou) [reply pkg/coscheduling/apis/podgroup/v1alpha1/doc.go:1]: But  I thinks this apis  are only for coscheduling

File: pkg/coscheduling/apis/podgroup/v1alpha1/types.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+type PodGroupSpec struct {
+	// MinMember defines the minimal number of members/tasks to run the pod group;
+	// if there's not enough resources to start all tasks, the scheduler
+	// will not start anyone.
+	MinMember uint32 `json:"minMember,omitempty"`
+
+	// MinResources defines the minimal resource of members/tasks to run the pod group;
+	// if there's not enough resources to start all tasks, the scheduler
+	// will not start anyone.
+	MinResources *v1.ResourceList `json:"minResources,omitempty"`
```
Comment (denkensk) [inline pkg/coscheduling/apis/podgroup/v1alpha1/types.go:88]: As comments before, MinResource was going to removed.
https://github.com/kubernetes-sigs/scheduler-plugins/pull/42#discussion_r478137253

File: pkg/coscheduling/apis/podgroup/v1alpha1/doc.go (hunk 1 above)
Comment (denkensk) [reply pkg/coscheduling/apis/podgroup/v1alpha1/doc.go:1]: PodGroup is a universal concept. It maybe used by other plugins.

File: pkg/coscheduling/apis/podgroup/v1alpha1/types.go (hunk 2 above)
Comment (cwdsuzhou) [reply pkg/coscheduling/apis/podgroup/v1alpha1/types.go:88]: https://github.com/kubernetes-sigs/scheduler-plugins/pull/42#discussion_r481790970, this it  the latest.

Comment (denkensk) [reply pkg/coscheduling/apis/podgroup/v1alpha1/types.go:88]: > I'm confused about how to set the MinResource. It's the total minimal resource of the PodGroup or the pod?
> yes
---
Yes？  the PodGroup or the pod？

**Comment by cwdsuzhou** [comment]
> We need to separate this PR to two @cwdsuzhou @Huang-Wei
> 
> 1. CRD Define + Controller
> 2. Use the PodGroup in the current coscheduling implemention (instead of deleteing the current coscheduling directly)

Too many difference between the existing version and this one. The existing version even can not pass the integration test.  If you do not like this, we would like move it to an independent `pkg` named `coscheduling-with-crd`. Actually we just use similar `interface`,  detailed implementations are absolutely different.

File: pkg/coscheduling/apis/podgroup/v1alpha1/types.go (hunk 2 above)
Comment (cwdsuzhou) [reply pkg/coscheduling/apis/podgroup/v1alpha1/types.go:88]: > > I'm confused about how to set the MinResource. It's the total minimal resource of the PodGroup or the pod?
> > yes
> 
> Yes？ the PodGroup or the pod？

We have discussed about this question offline. 

minimal resource of `PodGroup`

Comment (denkensk) [reply pkg/coscheduling/apis/podgroup/v1alpha1/types.go:88]: Ok. It's better to change like this:

// MinResources defines the minimal resource of members/tasks to run the pod group; ---> // MinResources defines the minimal resource to run the pod group;

**Comment by denkensk** [comment]
Please separate this PR to two firstly.
CRD Define + Controller
/hold

**Review by Huang-Wei — COMMENTED** [review]
Thanks for the PR @cwdsuzhou !

A high-level comment is that I'd still prefer not to mixin the implementation into this PR, i.e., it's ideal to have 2 PRs - one for CRD and controller, the 2nd one to include the implementation code. Additionally, the code structure needs to be tweaked a bit as we won't generate binaries for each plugin - all plugins will be compiled into particular type of binaries (scheduler, controller, admission, etc.).

With that said, here are my comments:

- Retitle the PR to "Implementation of coscheduling based on CRD [Part 1]"
- Move `cmd/main.go` and `cmd/main_test.go` to a new pkg `cmd/scheduler/`.
- Make `cmd/coscheduling/controller` a generic controller, i.e., move `cmd/coscheduler/controller/*` to `cmd/controller`
- Move `pkg/coscheduling/apis/podgroup` to `pkg/apis/podgroup` as `PodGroup` is a generic API, and hence may consumed by multiple plugins. It also follows the upstream code structure.
- `pkg/coscheduling/controller/*` should also go to `pkg/controller`, and named as `podgroup.go`.
- Move `pkg/coscheduling/generated` to `pkg/generated`
- IMO `pkg/coscheduling/scheduler/*`, along with `pkg/coscheduling/coscheduler.go` can be introduced in the next PR.
- Add tests to test PodGroup and controller logic only.
- Modify Dockerfile and Makefile based on latest codebase.

FYI: I've exercised some (not all) of the above comments, you can supplement based on it:
- https://github.com/Huang-Wei/scheduler-plugins/commit/13a7897a2513dec8cd9e7fc827999b4deca4d5cb (CRD definition and controller logic)
- https://github.com/Huang-Wei/scheduler-plugins/commit/b0ac55ec120aca31cfba85cca2a21e394158211e (codegen)

**Comment by cwdsuzhou** [comment]
> Thanks for the PR @cwdsuzhou !
> 
> A high-level comment is that I'd still prefer not to mixin the implementation into this PR, i.e., it's ideal to have 2 PRs - one for CRD and controller, the 2nd one to include the implementation code. Additionally, the code structure needs to be tweaked a bit as we won't generate binaries for each plugin - all plugins will be compiled into particular type of binaries (scheduler, controller, admission, etc.).
> 
> With that said, here are my comments:
> 
> * Retitle the PR to "Implementation of coscheduling based on CRD [Part 1]"
> * Move `cmd/main.go` and `cmd/main_test.go` to a new pkg `cmd/scheduler/`.
> * Make `cmd/coscheduling/controller` a generic controller, i.e., move `cmd/coscheduler/controller/*` to `cmd/controller`
> * Move `pkg/coscheduling/apis/podgroup` to `pkg/apis/podgroup` as `PodGroup` is a generic API, and hence may consumed by multiple plugins. It also follows the upstream code structure.
> * `pkg/coscheduling/controller/*` should also go to `pkg/controller`, and named as `podgroup.go`.
> * Move `pkg/coscheduling/generated` to `pkg/generated`
> * IMO `pkg/coscheduling/scheduler/*`, along with `pkg/coscheduling/coscheduler.go` can be introduced in the next PR.
> * Add tests to test PodGroup and controller logic only.
> * Modify Dockerfile and Makefile based on latest codebase.
> 
> FYI: I've exercised some (not all) of the above comments, you can supplement based on it:
> 
> * [Huang-Wei@13a7897](https://github.com/Huang-Wei/scheduler-plugins/commit/13a7897a2513dec8cd9e7fc827999b4deca4d5cb) (CRD definition and controller logic)
> * [Huang-Wei@b0ac55e](https://github.com/Huang-Wei/scheduler-plugins/commit/b0ac55ec120aca31cfba85cca2a21e394158211e) (codegen)

Thanks!

File: pkg/controller/podgroup_test.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	ctx := context.TODO()
+	podSuccess := v1.PodSucceeded
+	podFailed := v1.PodFailed
+	createTime := metav1.Time{Time: time.Now().Add(-72 * time.Hour)}
+	cases := []struct {
+		name               string
+		pgName             string
+		minMember          uint32
+		podNames           []string
+		podNextPhase       *v1.PodPhase
```
Comment (Huang-Wei) [inline pkg/controller/podgroup_test.go:31]: Can you make the type consistent? Make them all pointers, or all concrete types.

File: pkg/controller/podgroup_test.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			ps := pods(c.podNames, c.pgName, c.podPhase)
+			kubeClient := fake.NewSimpleClientset(ps[0], ps[1])
+			pg := pg(c.pgName, 2, c.previousPhase, c.podGroupCreateTime)
+			pgClient := pgfake.NewSimpleClientset(pg)
+
+			informerFactory := informers.NewSharedInformerFactory(kubeClient, controller.NoResyncPeriodFunc())
+			schedulingInformerFactory := pgformers.NewSharedInformerFactory(pgClient, controller.NoResyncPeriodFunc())
```
Comment (Huang-Wei) [inline pkg/controller/podgroup_test.go:121]: schedulingInformerFactory -> pgInformerFactory

File: pkg/controller/podgroup_test.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+			informerFactory.Start(ctx.Done())
+			schedulingInformerFactory.Start(ctx.Done())
+			if c.podNextPhase != nil {
+				ps := pods(c.podNames, c.pgName, *c.podNextPhase)
+				for _, p := range ps {
+					kubeClient.CoreV1().Pods(p.Namespace).UpdateStatus(ctx, p, metav1.UpdateOptions{})
+				}
+			}
+			go ctrl.Run(1, ctx.Done())
+			time.Sleep(200 * time.Millisecond)
```
Comment (Huang-Wei) [inline pkg/controller/podgroup_test.go:135]: Sleep() is a bad smell. If you want to wait for the status to get updated, use `wait.poll()` (or similar logic) to ensure that have happened.

File: pkg/controller/podgroup_test.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+			pg, err := pgClient.SchedulingV1alpha1().PodGroups("default").Get(ctx, c.pgName, metav1.GetOptions{})
+			if err != nil {
+				t.Fatal(err)
+			}
+			if pg.Status.Phase != c.desiredGroupPhase {
+				t.Fatalf("descired %v, get %v", c.desiredGroupPhase, pg.Status.Phase)
```
Comment (Huang-Wei) [inline pkg/controller/podgroup_test.go:142]:
```suggestion
				t.Fatalf("want %v, got %v", c.desiredGroupPhase, pg.Status.Phase)
```

**Review by Huang-Wei — COMMENTED** [review]
Just ignore the flaky test, #54 is created for tracking that.

File: cmd/controller/app/server.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+	informerFactory := informers.NewSharedInformerFactoryWithOptions(kubeClient, 0, informers.WithTweakListOptions(func(opt *metav1.ListOptions) {
+		opt.LabelSelector = util.PodGroupLabel
+	}))
+	podInformer := informerFactory.Core().V1().Pods()
+	ctrl := controller.NewPodGroupController(kubeClient, pgInformer, podInformer, pgClient)
+	informerFactory.Start(stopCh)
+	scheduleInformer.Start(stopCh)
+	run := func(ctx context.Context) {
+		ctrl.Run(10, ctx.Done())
```
[the remark this answers is not in the document]
Comment (cwdsuzhou) [reply cmd/controller/app/server.go:82]: I would make it configuration

File: pkg/controller/podgroup.go [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+	ctrl.pgQueue.Add(key)
+}
+
+// pgUpdated reacts to a PG update
+func (ctrl *PodGroupController) pgUpdated(old, new interface{}) {
+	ctrl.pgAdded(new)
+}
+
+// pgDelete reacts to a PG update
+func (ctrl *PodGroupController) pgDelete(new interface{}) {
```
[the remark this answers is not in the document]
Comment (cwdsuzhou) [reply pkg/controller/podgroup.go:132]: That is just for debug now. It is used for deleting cache in our initializing implementation.

Would remove it

File: pkg/controller/podgroup_test.go (hunk 5 above)
Comment (cwdsuzhou) [reply pkg/controller/podgroup_test.go:135]: Thanks

**Comment by cwdsuzhou** [comment]
PR updated according to your suggestions @Huang-Wei

File: cmd/controller/app/server.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+		if err != nil {
+			return err
+		}
+
+		// add a uniquifier so that two processes on the same host don't accidentally both become active
+		id = id + "_" + string(uuid.NewUUID())
+
+		rl, err := resourcelock.New("endpoints",
+			"kube-system",
+			"kube-scheduler",
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply cmd/controller/app/server.go:96]: This comment is still outstanding.

File: pkg/apis/podgroup/v1alpha1/types.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PodGroupList is a collection of pod groups.
+type PodGroupList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// items is the list of PodGroup
```
Comment (Huang-Wei) [inline pkg/apis/podgroup/v1alpha1/types.go:132]:
```suggestion
	// Items is the list of PodGroup
```

**Comment by Huang-Wei** [comment]
There is one comment unresolved: https://github.com/kubernetes-sigs/scheduler-plugins/pull/52#discussion_r491152457. LGTM otherwise.

@denkensk do you want to take another round of review?

File: cmd/controller/app/server.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (cwdsuzhou) [reply cmd/controller/app/server.go]: Done

File: pkg/apis/podgroup/v1alpha1/types.go (hunk 10 above)
Comment (cwdsuzhou) [reply pkg/apis/podgroup/v1alpha1/types.go]: updated

File: pkg/util/types.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+limitations under the License.
+*/
+
+package util
+
+import "fmt"
+
+const (
+	// PodGroupLabel is the default label of batch scheduler
+	PodGroupLabel = "podgroup.scheduling.sigs.k8s.io"
```
Comment (denkensk) [inline pkg/util/types.go:23]: podgroup.scheduling.sigs.k8s.io --> pod-group.scheduling.sigs.k8s.io

File: pkg/util/types.go [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+var (
+	// ErrorNotMatched means pod does not match batch scheduling
+	ErrorNotMatched = fmt.Errorf("not match batch scheduling")
```
Comment (denkensk) [inline pkg/util/types.go:28]: not match batch scheduling --> not match coscheduling

File: pkg/apis/podgroup/v1alpha1/types.go [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+	// The number of pods which reached phase Succeeded.
+	// +optional
+	Succeeded uint32 `json:"succeeded,omitempty"`
+
+	// The number of pods which reached phase Failed.
+	// +optional
+	Failed uint32 `json:"failed,omitempty"`
+
+	// ScheduleStartTime of the group
+	ScheduleStartTime metav1.Time `json:"scheduleStartTime,omitempty"`
```
Comment (denkensk) [inline pkg/apis/podgroup/v1alpha1/types.go:120]: ScheduleStartTime is the create time of PG ? 
If So, it's better to rename it to CreateTime?

File: pkg/util/types.go (hunk 12 above)
Comment (cwdsuzhou) [reply pkg/util/types.go:28]: Ok, thanks

File: pkg/apis/podgroup/v1alpha1/types.go (hunk 13 above)
Comment (cwdsuzhou) [reply pkg/apis/podgroup/v1alpha1/types.go:120]: This is field just record when the group starts scheduling

File: pkg/util/types.go (hunk 11 above)
Comment (cwdsuzhou) [reply pkg/util/types.go:23]: IMO， podgroup is alose OK. WDYT

Comment (denkensk) [reply pkg/util/types.go:23]: 1. `pod-group` conforms to the naming standard of k8s. We talk about it before.
2. Hope to be consistent with the current implementation of the naming

File: pkg/apis/podgroup/v1alpha1/types.go (hunk 13 above)
Comment (denkensk) [reply pkg/apis/podgroup/v1alpha1/types.go:120]: Thanks, It also need to add the field to record the create time of PG.

Comment (cwdsuzhou) [reply pkg/apis/podgroup/v1alpha1/types.go:120]: MetaData of CRD already includes this field!

File: pkg/util/types.go (hunk 11 above)
Comment (cwdsuzhou) [reply pkg/util/types.go:23]: Ok, done!

File: pkg/apis/podgroup/v1alpha1/types.go (hunk 13 above)
Comment (cwdsuzhou) [reply pkg/apis/podgroup/v1alpha1/types.go:120]: https://github.com/kubernetes-sigs/scheduler-plugins/pull/52/files#diff-23885a062c430d62621c160ab8aaaf08R66

Comment (denkensk) [reply pkg/apis/podgroup/v1alpha1/types.go:120]: Ok, Thanks

**Comment by Huang-Wei** [comment]
/approve

leaving /lgtm to @denkensk

**Comment by cwdsuzhou** [comment]
ping @denkensk

**Comment by cwdsuzhou** [comment]
@denkensk @Huang-Wei 

seems not work. May be you should left only /lgtm ?

**Pull request merged**